overlord/configstate/config/transaction.go

```diff
@@ -103,6 +104,49 @@ func (t *Transaction) Changes() []string {
 	return out
 }
 
+// shadowsVirtualConfig checks that the given subkeys/value does not
+// "block" the path to a virtual config with a non-map type. E.g. if
+// "network.netplan" is virtual it must be impossible to set
+// "network=false" or getting the document under "network" would be
+// wrong.
+func shadowsVirtualConfig(instanceName string, subkeys []string, value interface{}) error {
+	// maps never block the path
+	if v := reflect.ValueOf(value); v.Kind() == reflect.Map {
+		return nil
+	}
+
+	// helper to compare if two subkeys slices have a common prefix
+	keysHaveCommonPrefix := func(s, prefix []string) bool {
+		if len(prefix) > len(s) {
+			return false
+		}
+		for i := range prefix {
+			if prefix[i] != s[i] {
+				return false
+			}
+		}
+		return true
+	}
+
+	virtualMu.Lock()
+	km, ok := virtualMap[instanceName]
+	virtualMu.Unlock()
+	if ok {
+		for virtualKey := range km {
+			virtualSubKeys, err := ParseKey(virtualKey)
+			if err != nil {
+				return fmt.Errorf("internal error: invalid virtual configuration: %v", err)
+			}
+			if len(virtualSubKeys) > len(subkeys) && keysHaveCommonPrefix(virtualSubKeys, subkeys) {
```

Comment:
The first condition is already checked in `keysHaveCommonPrefix()`.

Reply:
Thanks! There is a subtle difference here (maybe that needs to be expressed better somehow). Suppose you have a virtual config: `key.nested.virtual` and you set `key.nested.virtual=1`. Then `len([key nested virtual]) > len(key nested virtual)` is not true and check for keysHaveCommonPrefix() is not called  which is what we want. It's fine to set a virtual config but not to override a virtual prefix with something that is not a map.

Reply:
You are right, I got confused by the order of the parameters and I now realize that inside the function we are actually doing `if len(virtualSubKeys) < len(subkeys) return false`. So, unless I'm mistaken again, the block inside this `if` condition is only taken if the two string arrays have the same length (and therefore, if they are exactly the same), right?

But if this is the case, since this is the only place where we use this `keysHaveCommonPrefix()` function, can't we just remove it and instead compare the two arrays for equality?